### DISCUSSION TITLE
Pushing docker image to GPR: manifest invalid: Only schema version 2 is supported

### DISCUSSION THREAD

**Discussion by WyriHaximus** [community/community#25315] [Packages]
<p>Since about an hour I’m getting the following error when I try to push a docker image tp GPR:</p>
<pre><code>[DEPRECATION NOTICE] registry v2 schema1 support will be removed in an upcoming release. Please contact admins of the docker.pkg.github.com registry NOW to avoid future disruption. More information at https://docs.docker.com/registry/spec/deprecated-schema-v1/
manifest invalid: Only schema version 2 is supported
</code></pre>
<p>Two hours ago everything work fine, the Dockerfile hasn’t changed, neither did the upstream image. Did Github change something an hour ago?</p>

**Comment by jmahowald** [comment]
<p>I also had things working, then they started failing.</p>
<p>From here there did seem to be someone who found tagging with the git sha was problematic (too many characters).</p>
<p><a href="https://github.community/t/cant-push-to-github-docker-registry-from-actions/16161">https://github.community/t5/GitHub-Actions/Can-t-push-to-GitHub-Docker-Registry-from-Actions/td-p/30665</a></p>

**Comment by WyriHaximus** [comment]
<p>Tried it with the following tags so I doubt it’s the tag langth: <code>nieuwste</code>, <code>laatste</code>, <code>la</code></p>
<p>Also interesting detail is that I’m building the image I’m building this image on publicly on Github Actions and push it to Docker Hub. (Frequently, in this case the base image was less then a day old.)</p>

**Comment by thepont** [comment]
<p>I have the same issue pushing to the Github repository but via CircleCI, can you update the version of docker client your using it the Github action for pushing… ( This may be a case of just updating the version of the GitHub action you’re using for pubishing)</p>
<p>I think its just because the old API has been deprecated and we are still pushing using older versions of the api ?</p>
<p>I updated my docker image used to publish my image to Docker version 19.03.5, build 633a0ea838 with no success, build is still failing with the metadata error.</p>

**Comment by zymartinez** [comment]
<p>I tried your suggestion by trying to push outside of Github Actions. I’m currently on Docker version 19.03.2, build 6a30dfc on mac and I’m still getting the manifest invalid error. This may be caused by a different issue?</p>

**Comment by thepont** [comment]
<p>Sorry, I just got mine building with another version and had the same issue again… Are people using the github private repo ?</p>
<p>I also don’t know if it’s worth noting that the docker manual says :</p>
<p>&gt; When pushing images, clients which support the new manifest format should first construct a manifest in</p>
<p>&gt; the new format. If uploading this manifest fails, presumably because the registry only supports the old</p>
<p>&gt; format, the client may fall back to uploading a manifest in the old format.</p>
<p>This makes me think that it may be failing to upload using v2 then reverting to v1 where it then gets this error back from the server.</p>

**Comment by WyriHaximus** [comment]
<p>Yes I’m trying to push from a private repo to it’s private repo. This fallback to v1 could be it, there is a noticeable delay after the last layer is pushed before the error shows up.</p>
<p>Nevermind that, for both success and failure the delay is 6 seconds. Success gives you the latest digest SHA And failure that error.</p>

**Comment by thepont** [comment]
<p>FYI, I sent a message detailing the issue github developer support, who have escalated the issue to the product team with the link to this community post.</p>
<p>I’ll update if I hear back from them.</p>
<p><strong>Update, I figured out my issue:</strong></p>
<p>So nested in the logs ( journalctl ) is the actual issue:</p>
<pre><code>Nov 20 09:42:08 blue-pont dockerd[427]: time="2019-11-20T09:42:08.376878027+11:00" level=warning msg="failed to upload schema2 manifest: unauthorized: This operation would exceed the storage allotment for this account. - falling back to schema1"
Nov 20 09:42:08 blue-pont dockerd[427]: time="2019-11-20T09:42:08.376989057+11:00" level=warning msg="[DEPRECATION NOTICE] registry v2 schema1 support will be removed in an upcoming release. Please contact admins of the docker.pkg.github.com registry NOW to avoid future disruption."
Nov 20 09:42:10 blue-pont dockerd[427]: time="2019-11-20T09:42:10.260794961+11:00" level=error msg="Not continuing with push after error: manifest invalid: Only schema version 2 is supported"
</code></pre>
<p>Seems our account is over the storage allocation ( I’m not the certain what this limit is ) and therefore we are no longer able to upload our images.</p>
<p>It is attempting to upload using schema 2 then failing and falling back, docker does not know how to handle the error from the github private repo and is therefor falling back in a reall confusing manner.</p>
<p>So i am not certain if we can delete older packages to free up space or what the limit is.</p>
<p>I used journalctl | grep dockerd (since I aways forget the flags) to get the full logs on Arch Linux, this will be different on OSX (<a href="https://docs.docker.com/docker-for-mac/troubleshoot/" rel="nofollow noopener">https://docs.docker.com/docker-for-mac/troubleshoot/</a>) or windows (<a href="https://docs.docker.com/v17.12/docker-for-windows/troubleshoot/" rel="nofollow noopener">https://docs.docker.com/v17.12/docker-for-windows/troubleshoot/</a>).</p>
<p>But Docker will log the reason it falled back to the version 1 schema in its logs.</p>

**Comment by thepont** [comment]
<p>So nested in the logs ( journalctl ) is the actual issue:</p>
<pre><code>Nov 20 09:42:08 blue-pont dockerd[427]: time="2019-11-20T09:42:08.376878027+11:00" level=warning msg="failed to upload schema2 manifest: unauthorized: This operation would exceed the storage allotment for this account. - falling back to schema1"
Nov 20 09:42:08 blue-pont dockerd[427]: time="2019-11-20T09:42:08.376989057+11:00" level=warning msg="[DEPRECATION NOTICE] registry v2 schema1 support will be removed in an upcoming release. Please contact admins of the docker.pkg.github.com registry NOW to avoid future disruption."
Nov 20 09:42:10 blue-pont dockerd[427]: time="2019-11-20T09:42:10.260794961+11:00" level=error msg="Not continuing with push after error: manifest invalid: Only schema version 2 is supported"
</code></pre>
<p>Seems our account is over the storage allocation ( I’m not the certain what this limit is ) and therefore we are no longer able to upload our images.</p>
<p>It is attempting to upload using schema 2 then failing and falling back, docker does not know how to handle the error from the github private repo and is therefor falling back in a reall confusing manner.</p>
<p>So i am not certain if we can delete older packages to free up space or what the limit is.</p>
<p>I used journalctl | grep dockerd (since I aways forget the flags) to get the full logs on Arch Linux, this will be different on OSX (<a href="https://docs.docker.com/docker-for-mac/troubleshoot/" rel="nofollow noopener">https://docs.docker.com/docker-for-mac/troubleshoot/</a>) or windows (<a href="https://docs.docker.com/v17.12/docker-for-windows/troubleshoot/" rel="nofollow noopener">https://docs.docker.com/v17.12/docker-for-windows/troubleshoot/</a>).</p>
<p>But Docker will log the reason it falled back to the version 1 schema in its logs.</p>

**Comment by aledalgrande** [comment]
<p>We are seeing the same error and we have fixed tag length. Started after a few days, it was working before. Hopefully some staff member can write here to give more information.</p>

**Comment by aledalgrande** [comment]
<p>A related question: if the space is limited, can we automatically remove old images via some command/API?</p>

**Comment by WyriHaximus** [comment]
<p>Yes we can, I already have a work in progress action to remove package versions: <a href="https://github.com/WyriHaximus/github-action-delete-package" rel="nofollow noopener">https://github.com/WyriHaximus/github-action-delete-package</a></p>
<p>Which is based on this bit of documentation: <a href="https://help.github.com/en/github/managing-packages-with-github-packages/deleting-a-package#deleting-a-version-of-a-private-package" rel="nofollow noopener">https://help.github.com/en/github/managing-packages-with-github-packages/deleting-a-package#deleting-a-version-of-a-private-package</a></p>
<p>The thing is I’m trying to get the list of this packages with this GraphQL query but it’s giving me an unexpected not found error:</p>
<pre><code>query {
  repository(owner:"OWNER", name:"REPOSITORY") {
    registryPackages(last:20) {
      edges {
        node {
          id,
          name
        }
      }
    }
  }
}
</code></pre>
<p>Glueing these two together and get them to work should make you able to delete old docker image tags.</p>

**Comment by aledalgrande** [comment]
<p>UPDATE: When we increased our total quota for the registry the action started working.</p>
<p>Also looking into that delete API, but we have a different approach because we don’t want to delete last n, we want to delete all images related to a tag too. Pretty simple to call the API once you get your head around it.</p>
<p>I find compounding actions with <code>uses</code> increases the total run time (have to download the image and run it), so unless the task is complex, it’s better to add a script/step to do it IMO. Also the least dependencies the better.</p>

**Comment by thepont** [comment, marked as answer]
<p>I might be stating things that people already know, But talking to github support they advised that the default storage for Github team is: </p>
<p>GitHub Team</p>
<ul>
<li>Storage: 2GB</li>
<li>Data Transfer: 10GB</li>
</ul>
<p>Also that the organisation storage can be checked by viewing: <a href="https://github.com/organizations/ORG/settings/billing" rel="nofollow noopener">https://github.com/organizations/ORG/settings/billing</a></p>
<p>The following works for me, to list versions of packages on a project.</p>
<pre><code>curl -X POST \
    -H "Accept: application/vnd.github.package-deletes-preview+json" \
    -H "Authorization: bearer TOKEN" \
    -d '{"query":"query {repository(owner:\"ORG\", name:\"REPOSITORY\") {registryPackages(last:20) {edges{node{id, name, versions(last:20){edges {node {id}}}}}}}}"}' \
    https://api.github.com/graphql
</code></pre>

**Comment by aledalgrande** [comment]
<p>One thing you have to be careful with, we just discovered, is not deleting all the versions for a package; Github at that point will prevent you from publishing <em>any</em> version under that name and will return this error:</p>
<pre><code>blob upload invalid: Docker image with name "&lt;image name&gt;" is already associated with another repository under owner "&lt;account&gt;"
</code></pre>
<p>Still waiting for an answer from support to understand how to fix it.</p>

**Comment by WyriHaximus** [comment]
<p><a class="mention" href="/thepont">@thepont</a> Yeah that fixes it. Would be nice to get a more descriptive error message in the long run tbh. Thank you for fixing the GraphQL query as well, they will get me to something working this weekend 🎉.</p>
<p><a class="mention" href="/aledalgrande">@aledalgrande</a> That’s good to know! Wanted to do a latest tag for master builds and SHA or tag title for tagged releases, but sounds like the latest tag for master builds isn’t an option then :(. The reason I’m also doing an Action is to both give back and make it easier to compose actions between my projects. I’m intending to build an auto clean up of old image tags that aren’t relevant anymore but are still taking up space. (Also where I can I’m building Actions directly on alpine with a quick package install keeping that build time as low as possible.)</p>
<p><a class="mention" href="/thepont">@thepont</a> <a class="mention" href="/aledalgrande">@aledalgrande</a> Thank you both for all the help on this. I’ve accepted <a class="mention" href="/thepont">@thepont</a>’s post because it’s a nice concise summary, but want to let you know you both been very helpful 👍.</p>

**Comment by patped** [comment]
<p>Any update from support on this error?</p>

**Comment by saesh** [comment]
<p>Did you ever get an answer? I am having the exact same problem now.</p>

**Comment by kevin-beatdapp** [comment]
<p>I’m experiencing the same issue now, were you able to find a solution?</p>

**Comment by dieswaytoofast** [comment]
<p>How did any of you delete a docker package?<br>
when i try to do so, it sez. “Version deletion is currently unsupported for docker”</p>
